Jv integration tests for runtime config

JoukoVirtanen opened the pull request:
## Description

Integration tests for runtime configuration. The tests create various versions of the runtime configuration file and check that the configuration introspection endpoint returns the correct expected configuration.

## Checklist
- [x] Investigated and inspected CI test results
- [ ] Updated documentation accordingly

**Automated testing**
  - [ ] Added unit tests
  - [x] Added integration tests
  - [ ] Added regression tests

If any of these don't apply, please comment below.

## Testing Performed

Ran the integration tests locally 2,000 times without a failure using the following script
```
#!/usr/bin/env bash
set -eou pipefail

for ((i = 0; i < 2000; i=i+1)); do
        echo "i= $i"
        sudo -E make -C integration-tests TestRuntimeConfigFile
done
```

JoukoVirtanen commented on collector/lib/ConfigLoader.cpp:
```diff
@@ -47,6 +47,10 @@ bool ConfigLoader::LoadConfiguration(CollectorConfig& config) {
+  } catch (const YAML::Exception& e) {
+    CLOG(ERROR) << "Failed to parse the configuration file: " << config_file << ". Error: " << e.what();
+  } catch (const std::exception& e) {
+    CLOG(ERROR) << "An unknown error occured while loading the configuration file: " << config_file << ". Error: " << e.what();
```
Remove

JoukoVirtanen replied:
Removed

Molter73 commented on collector/lib/ConfigLoader.cpp:
```diff
@@ -55,12 +59,13 @@ bool ConfigLoader::LoadConfiguration(CollectorConfig& config) {
+
   YAML::Node networking = node["networking"];
-  if (!networking) {
+  if (!node["networking"] || node["networking"].IsNull()) {
```
```suggestion
  if (!networking || networking.IsNull()) {
```

Molter73 commented on collector/test/ConfigLoaderTest.cpp:
```diff
@@ -67,12 +67,20 @@ TEST(CollectorConfigTest, TestYamlConfigToConfigInvalid) {
 }
 
 TEST(CollectorConfigTest, TestYamlConfigToConfigEmpty) {
```
Please update this test name.

JoukoVirtanen replied:
Done

Molter73 commented on integration-tests/pkg/collector/collector_docker.go:
```diff
@@ -41,6 +41,7 @@ func NewDockerCollectorManager(e executor.Executor, name string) *DockerCollecto
 		"/host/etc:ro":              "/etc",
 		"/host/usr/lib:ro":          "/usr/lib",
 		"/host/sys/kernel/debug:ro": "/sys/kernel/debug",
+		"/etc/stackrox:ro":          "/tmp",
```
Consider creating a separate directory for this like `/tmp/collector-test` and use that instead of the full /tmp directory.

JoukoVirtanen replied:
Done

Molter73 commented on integration-tests/pkg/collector/introspection.go:
```diff
@@ -23,3 +30,50 @@ func (k *K8sCollectorManager) IntrospectionQuery(endpoint string) ([]byte, error
+}
+
+func AssertExternalIps(t *testing.T, enable bool, collectorIP string) {
+	AssertRepeated(t, func() bool {
+		body := QueryConfig(t, collectorIP)
+		var response types.RuntimeConfig
+		err := json.Unmarshal(body, &response)
+		assert.NoError(t, err)
+
+		return response.Networking.ExternalIps.Enable == enable
+	})
+}
+
+func AssertNoRuntimeConfig(t *testing.T, collectorIP string) {
+	AssertRepeated(t, func() bool {
+		body := QueryConfig(t, collectorIP)
+		return strings.TrimSpace(string(body)) == "{}"
+	})
+}
+
+// TODO: This should be in its own package
+func AssertRepeated(t *testing.T, condition func() bool) {
+	tick := time.NewTicker(1 * time.Second)
+	timer := time.After(3 * time.Minute)
+
+	for {
+		select {
+		case <-tick.C:
+			if condition() {
+				// Condition has been met
+				return
+			}
+
+		case <-timer:
+			// TODO: This message should be passed in rather than hard coded here
+			t.Log("Timeout reached: Runtime configuration was not updated")
+			t.FailNow()
+		}
+	}
+}
```
We probably want to move all assertions to a separate `pkg/assert` or similar.

JoukoVirtanen replied:
Done

Molter73 commented on integration-tests/pkg/collector/introspection.go:
```diff
@@ -23,3 +30,50 @@ func (k *K8sCollectorManager) IntrospectionQuery(endpoint string) ([]byte, error
 	return body, err
 }
+
+func QueryConfig(t *testing.T, collectorIP string) []byte {
+	log.Info("Querying: /state/config")
+	body, err := IntrospectionQuery(collectorIP, "/state/config")
+	assert.NoError(t, err)
+	log.Info("Response: %q", body)
+	return body
+}
```
I'm not 100% sold this is something we need, from a test we could just do:
```go
body, err := IntrospectionQuery(s.Collector().IP(), "/state/config")
s.Require().NoError(err)
```
This is simple enough and you can add the logging if you deem it necessary.

JoukoVirtanen replied:
Done

Molter73 reviewed (commented):
I'm less than half way through, I'll finish the review tomorrow.

JoukoVirtanen commented on integration-tests/pkg/mock_sensor/expect_conn.go:
```diff
@@ -33,7 +33,7 @@ loop:
 			// we know they don't match at this point, but by using
 			// ElementsMatch we get much better logging about the differences
-			return assert.ElementsMatch(t, expected, s.Connections(containerID), "timed out waiting for networks")
+			return assert.ElementsMatch(t, expected, s.Connections(containerID), "timed out waiting for network connections")
```
The changes in this file were made here https://github.com/stackrox/collector/pull/1902/files

JoukoVirtanen commented on integration-tests/pkg/mock_sensor/server.go:
```diff
@@ -32,7 +32,6 @@ const (
 // us to use any comparable type as the key)
 type ProcessMap map[types.ProcessInfo]interface{}
 type LineageMap map[types.ProcessLineage]interface{}
-type ConnMap map[types.NetworkInfo]interface{}
```
The changes in this file were made here https://github.com/stackrox/collector/pull/1902/files

Molter73 commented on integration-tests/suites/runtime_config_file.go:
```diff
@@ -0,0 +1,172 @@
+}
+
+func (s *RuntimeConfigFileTestSuite) setExternalIpsEnable(runtimeConfigFile string, enable bool) {
+	configStr := "networking:\n  externalIps:\n    enable: " + strconv.FormatBool(enable)
```
```suggestion
	configStr := `networking:
  externalIps:
    enable: ` + strconv.FormatBool(enable)
```

Or even better, use the `RuntimeConfig` type defined in `integration-tests/pkg/types/runtime_config.go` and serialize it to yaml.

Molter73 commented on integration-tests/suites/runtime_config_file.go:
```diff
@@ -0,0 +1,172 @@
+	externalIp := "8.8.8.8"
+	serverPort := 53
+	externalUrl := fmt.Sprintf("http://%s:%d", externalIp, serverPort)
+	image_store := config.Images()
+	containerID, err := s.Executor().StartContainer(
+		config.ContainerStartConfig{
+			Name:    "external-connection",
+			Image:   image_store.QaImageByKey("qa-alpine-curl"),
+			Command: []string{"sh", "-c", "while true; do curl " + externalUrl + "; sleep 1; done"},
+		})
```
[nit]
```suggestion
	containerID, err := s.Executor().StartContainer(
		config.ContainerStartConfig{
			Name:    "external-connection",
			Image:   config.Images().QaImageByKey("qa-alpine-curl"),
			Command: []string{"sh", "-c", "while true; do curl " + externalUrl + "; sleep 1; done"},
		})
```

Molter73 commented on integration-tests/suites/runtime_config_file.go:
```diff
@@ -0,0 +1,172 @@
+	assert.AssertExternalIps(s.T(), true, collectorIP)
+	// When external IPs is enabled the normalized connection is reported as inactive and the unnormalized connection
+	// is reported as active
+	expectedConnections = append(expectedConnections, activeUnnormalizedConnection, inactiveNormalizedConnection)
```
Shouldn't we be cleaning `expectedConnections` between each test? Isn't this just saying "you should see the previous ones and these ones"? How does that work when we turn external IPs back off?

JoukoVirtanen replied:
I don't think we should clean `expectConnections` the mock server does not clear out the connections that it sees after asserts, so `expectConnections` should not be cleared out either.

> Isn't this just saying "you should see the previous ones and these ones"? 
Yes

 > How does that work when we turn external IPs back off?
 When external IPs is turned back off the connection with the unnormalized connection is reported closed and the normalized connection is reported open. We get two reported connection events in that case.

Molter73 commented on integration-tests/suites/runtime_config_file.go:
```diff
@@ -0,0 +1,172 @@
+
+	s.deleteFile(runtimeConfigFile)
+	assert.AssertNoRuntimeConfig(s.T(), collectorIP)
+	// When the config file is deleted the config is reset and external IPs should continue to be disabled. There is no
+	// change in the expected connections.
+	common.Sleep(3 * time.Second) // Sleep so that collector has a chance to report connections
+	connectionSuccess = s.Sensor().ExpectSameElementsConnections(s.T(), clientContainer, 10*time.Second, expectedConnections...)
+	s.Require().True(connectionSuccess)
+
+	s.setExternalIpsEnable(runtimeConfigFile, true)
+	assert.AssertExternalIps(s.T(), true, collectorIP)
+	// The runtime config file is recreated with external IPs enabled. The unnormalized connection is reported as being active
+	// and the normalized connection is reported as being inactive.
+	expectedConnections = append(expectedConnections, activeUnnormalizedConnection, inactiveNormalizedConnection)
+	connectionSuccess = s.Sensor().ExpectSameElementsConnections(s.T(), clientContainer, 10*time.Second, expectedConnections...)
+	s.Require().True(connectionSuccess)
+
+	s.deleteFile(runtimeConfigFile)
+	assert.AssertNoRuntimeConfig(s.T(), collectorIP)
+	// The runtime config file is deleted again. This time there should be a change as external IPs transitions from being
+	// enabled to disabled. The normalized connection should be active and the unnormalized connection shoul be inactive.
+	expectedConnections = append(expectedConnections, activeNormalizedConnection, inactiveUnnormalizedConnection)
+	connectionSuccess = s.Sensor().ExpectSameElementsConnections(s.T(), clientContainer, 10*time.Second, expectedConnections...)
+	s.Require().True(connectionSuccess)
+
+	s.setExternalIpsEnable(runtimeConfigFile, false)
+	assert.AssertExternalIps(s.T(), false, collectorIP)
+	// The runtime config file is created one more time, this time with external IPs disabled. Since there is no change in the
+	// state, there should not be a change in the connections reported
+	common.Sleep(3 * time.Second) // Sleep so that collector has a chance to report connections
+	connectionSuccess = s.Sensor().ExpectSameElementsConnections(s.T(), clientContainer, 10*time.Second, expectedConnections...)
+	s.Require().True(connectionSuccess)
+
+	invalidConfig := "asdf"
+	s.setRuntimeConfig(runtimeConfigFile, invalidConfig)
+	assert.AssertExternalIps(s.T(), false, collectorIP)
+	// Testing an invalid configuration. There should not be a change in the configuration or reported connections
+	common.Sleep(3 * time.Second) // Sleep so that collector has a chance to report connections
+	connectionSuccess = s.Sensor().ExpectSameElementsConnections(s.T(), clientContainer, 10*time.Second, expectedConnections...)
+	s.Require().True(connectionSuccess)
```
Couldn't this be split into multiple test cases? Reviewing this test is borderline impossible.

JoukoVirtanen replied:
It could be, but I think this is better. This tests for case when external IPs is enabled after the config file is deleted as well as the case when external IPs is disabled after the config file is deleted. It tests for deleting the config file when external IPs is enabled and it tests for deleting the config file with external IPs is disabled. These cases could be split into different tests, but it would take longer to run. Would better comments make it easier to understand what is going on in the test?

Molter73 replied:
> Couldn't this be split into multiple test cases? Reviewing this test is borderline impossible.

It's not a problem if they take a couple more seconds, the problem is that you have 8, highly condensed, nearly identical blocks of code. This makes it really hard to understand what is actually changing from one block to the next, because I can't see what the difference is. This, in turn, makes it really hard for me to give meaningful feedback on the test, because I can't understand it.

The intent of the test should be obvious from a glance at the code.

> Would better comments make it easier to understand what is going on in the test?

If anything, the existing comments are making the code even denser and harder to understand. Try doing shorter comments at the start of each block, don't go into deep details of what the block is doing, just give a hint of what it'll do and try to read the block of code from the perspective of a person that is going through it for the first time, ask yourself:
- Is it easy to see what has changed in this part of the test?
- Can I easily spot the intent of the block?
- Is it distinct enough from a previous, similar block?

JoukoVirtanen replied:
I have split the test into three tests.

JoukoVirtanen commented on integration-tests/pkg/assert/assert.go:
```diff
@@ -0,0 +1,53 @@
+	"github.com/stackrox/collector/integration-tests/pkg/types"
+)
+
+func AssertExternalIps(t *testing.T, enable bool, collectorIP string) {
+	AssertRepeated(t, func() bool {
+		body, err := collector.IntrospectionQuery(collectorIP, "/state/config")
```
I think it can be done in another PR. Maybe we should have an `Introspection` struct that could be a member of a `Collector` interface. The introspection object would get the IP address from the collector object.

Molter73 reviewed (commented):
Thanks for splitting the tests up, as silly as it may sound, it's a lot easier to understand what is going on now.

The comments I left now are mostly on moving some comments around to leave the code blocks on their own. This should help reduce clutter a bit more, since the description of what will be done is provided up front and the code is left as clean as possible.

JoukoVirtanen commented:
CI results

Run 1
Many Konflux integration tests failed. rhcos failed with 
```
=== RUN   TestProcfsScraperDisableFeatureFlag/TestProcfsScraper
    expect_conn.go:181: 
        	Error Trace:	/tests/pkg/mock_sensor/expect_conn.go:181
        	            				/tests/suites/procfs_scraper.go:72
        	Error:      	elements differ
        	            	
        	            	extra elements in list A:
        	            	([]interface ***) (len=1) ***
        	            	 (types.EndpointInfo) ***
        	            	  Protocol: (string) (len=15) "L4_PROTOCOL_TCP",
        	            	  Address: (types.ListenAddress) ***
        	            	   AddressData: (string) (len=4) "\x00\x00\x00\x00",
        	            	   Port: (int) 80,
        	            	   IpNetwork: (string) (len=5) "\x00\x00\x00\x00 "
        	            	  ***,
        	            	  CloseTimestamp: (string) (len=5) "<nil>",
        	            	  Originator: (types.ProcessOriginator) ***
        	            	   ProcessName: (string) "",
        	            	   ProcessExecFilePath: (string) "",
        	            	   ProcessArgs: (string) ""
        	            	  ***
        	            	 ***
        	            	***
        	            	
        	            	
        	            	listA:
        	            	([]types.EndpointInfo) (len=1) ***
        	            	 (types.EndpointInfo) ***
        	            	  Protocol: (string) (len=15) "L4_PROTOCOL_TCP",
        	            	  Address: (types.ListenAddress) ***
        	            	   AddressData: (string) (len=4) "\x00\x00\x00\x00",
        	            	   Port: (int) 80,
        	            	   IpNetwork: (string) (len=5) "\x00\x00\x00\x00 "
        	            	  ***,
        	            	  CloseTimestamp: (string) (len=5) "<nil>",
        	            	  Originator: (types.ProcessOriginator) ***
        	            	   ProcessName: (string) "",
        	            	   ProcessExecFilePath: (string) "",
        	            	   ProcessArgs: (string) ""
        	            	  ***
        	            	 ***
        	            	***
        	            	
        	            	
        	            	listB:
        	            	([]types.EndpointInfo) ***
        	            	***
        	Test:       	TestProcfsScraperDisableFeatureFlag/TestProcfsScraper
        	Messages:   	timed out waiting for endpoints
```
The other konflux integration tests failed with either problems pulling the image. Both konflux and non-konflux cos integration tests failed with verifier errors.

Molter73 commented on integration-tests/suites/runtime_config_file.go:
```diff
@@ -0,0 +1,197 @@
+	// The runtime config file was deleted before starting collector so there should not be any config
+	assert.AssertNoRuntimeConfig(s.T(), collectorIP)
+	// Since there is no config the default is used, which means external IPs is disabled and we should
+	// expect a normalized connection
```
This was not applied to all tests.

JoukoVirtanen replied:
It has now been applied to all tests.

Molter73 reviewed (commented):
Changes look ok, but #1902 still has comments on it that haven't been addressed, if your intent is to merge the changes from that PR with this one, please close it so I can move my comments to this PR.

Stringy commented on integration-tests/pkg/assert/assert.go:
```diff
@@ -0,0 +1,53 @@
+}
+
+func AssertRepeated(t *testing.T, condition func() bool) {
+	tick := time.NewTicker(1 * time.Second)
+	timer := time.After(3 * time.Minute)
```
Can these timeouts be configurable via some parameters?

JoukoVirtanen replied:
Done

JoukoVirtanen commented:
> Changes look ok, but #1902 still has comments on it that haven't been addressed, if your intent is to merge the changes from that PR with this one, please close it so I can move my comments to this PR.

I have addressed the PR review comments in the other PR.

Molter73 commented on integration-tests/pkg/assert/assert.go:
```diff
@@ -0,0 +1,93 @@
+
+		case <-timer:
+			// TODO: This message should be passed in rather than hard coded here
+			t.Log("Timeout reached: Runtime configuration was not updated")
```
We probably want to use our log package for consistency
```suggestion
			log.Error("Timeout reached: Runtime configuration was not updated")
```

Molter73 commented on integration-tests/pkg/mock_sensor/server.go:
```diff
@@ -432,11 +432,10 @@ func (m *MockSensor) pushConnection(containerID string, connection *sensorAPI.Ne
 
-	if connections, ok := m.connections[containerID]; ok {
-		connections[conn] = true
+	if _, ok := m.connections[containerID]; ok {
+		m.connections[containerID] = append(m.connections[containerID], conn)
```
[nit] Not 100% sure this will work.
```suggestion
	if c, ok := m.connections[containerID]; ok {
		m.connections[containerID] = append(c, conn)
```

Molter73 reviewed (commented):
LGTM, but holding approval until #1902 is merged, so this can be rebased on top of that.

Molter73 commented on integration-tests/pkg/assert/assert.go:
```diff
@@ -1,13 +1,64 @@
+			}
+
+		case <-timer:
+			// TODO: This message should be passed in rather than hard coded here
+			log.Error("Timeout reached: Runtime configuration was not updated")
```
This `TODO` should be relatively simple to solve, add a format and args parameter to the function like this:
```go
func AssertRepeated(t *testing.T, tickTime time.Duration, timeout time.Duration, condition func() bool, msg string, args ...interface{}) {
```
Then here you can just call `log.Error` like this:
```go
log.Error("Timeout reached: " + msg, args...)
```

If this ends up being too verbose on the call sites with the function being defined inline, you can always define the function and assign it to a variable, then pass that in.

Molter73 reviewed (approved):
LGTM!

Just a small comment on a TODO that I think we could solve quickly before merging, but we can also address it on the next PR that needs to use the helper function.

Stringy commented on integration-tests/pkg/assert/assert.go:
```diff
@@ -1,13 +1,67 @@
+	runtimeConfigErrorMsg = "Runtime configuration was not updated"
 )
 
+func AssertExternalIps(t *testing.T, enable bool, collectorIP string) {
```
[nit] should probably be `enabled` rather than `enable` (the latter implies we're turning it on)

JoukoVirtanen replied:
Done

Stringy commented on integration-tests/pkg/types/runtime_config.go:
```diff
@@ -0,0 +1,13 @@
+		ExternalIps struct {
+			Enable bool
+		}
+	}
```
just a note that we'll need to add `PerContainerRateLimit` (probably in a follow up PR)

JoukoVirtanen replied:
We should not be defining a RuntimeConfig struct here. We should be importing the go code generated from the protobuf definitions. That should be done in a different PR.

Stringy commented on integration-tests/suites/runtime_config_file.go:
```diff
@@ -0,0 +1,195 @@
+}
+
+func (s *RuntimeConfigFileTestSuite) createDirectory(dir string) {
+	cmd := "mkdir " + dir
+	s.execShellCommand(cmd)
```
Can't we do this via the standard library? same for deleteFile, setRuntimeConfig

JoukoVirtanen replied:
This way we get the logging that comes with `execShellCommand`. Also if there are changes to how commands need to be executed, using the standard library might not be the best method. E.g we might want to run the commands in a VM, locally, in a container, or somewhere else and changes to those things might result in the standard library not working. However, you did remind me that this is not the correct place for these functions and I have moved them to `base.go`.

Stringy replied:
We've gone to great lengths to remove the various contexts in which things run, and today we're always running on the same host with the test containers so I don't think we should care about how commands run, or whether to restrict ourselves to using the shell instead of the stdlib. If this ever changes, then we'd have to overhaul the way the tests work again, so a few file system operations would be the least of our worries.

If these specific operations need to run in various contexts, then `execShellCommand` is still inappropriate because it is intended for running a local shell command (via `exec.Command`) and we'd have to change these functions anyway. 

It's just that a shell command feels very expensive for an operation as simple as making a directory or deleting a file, and if we need logging then we can add logging here.

Stringy replied:
Also, we would have a better handle on the errors if we do it via the stdlib. We wouldn't need to worry about failures in exec'ing the shell or the command, for example.

we're also just suppressing all errors at the moment -- we should probably return them to the caller

JoukoVirtanen replied:
Done

Stringy commented on integration-tests/suites/runtime_config_file.go:
```diff
@@ -0,0 +1,195 @@
+}
+
+func (s *RuntimeConfigFileTestSuite) setExternalIpsEnable(runtimeConfigFile string, enable bool) {
+	configStr := `networking:
+  externalIps:
+    enable: ` + strconv.FormatBool(enable)
+	s.setRuntimeConfig(runtimeConfigFile, configStr)
```
It would be better for to construct the object and serialize it to a string, and in fact it may be best to have the caller pass the structure directly to setRuntimeConfig

```
func (s *RuntimeConfigFileTestSuite) setRuntimeConfig(path string, config types.RuntimeConfig) error
```

JoukoVirtanen replied:
Done

Stringy commented on integration-tests/suites/runtime_config_file.go:
```diff
@@ -0,0 +1,189 @@
+	s.execShellCommand(cmd)
+}
+
+func (s *RuntimeConfigFileTestSuite) setExternalIpsEnable(runtimeConfigFile string, enable bool) {
```
This might sound like a nit, but from a separation of concerns perspective, this function shouldn't do the marshalling - it should be handled by `setRuntimeConfig` 

This function should just construct `types.RuntimeConfig` object and pass it to `setRuntimeConfig`. Otherwise, when we inevitably come to adding another field to these tests, we'll have to duplicate the marshaling logic.

JoukoVirtanen replied:
Done

JoukoVirtanen commented on integration-tests/pkg/types/runtime_config.go:
```diff
@@ -0,0 +1,27 @@
+	return n.Networking.ExternalIps.Enable == other.Networking.ExternalIps.Enable
+}
+
+func (s *RuntimeConfig) GetRuntimeConfigStr() (string, error) {
```
```suggestion
func (n *RuntimeConfig) GetRuntimeConfigStr() (string, error) {
```

JoukoVirtanen commented on integration-tests/suites/runtime_config_file.go:
```diff
@@ -0,0 +1,192 @@
+
+func (s *RuntimeConfigFileTestSuite) setRuntimeConfig(runtimeConfigFile string, configStr string) {
+	cmd := "echo '" + configStr + "' > " + runtimeConfigFile
+	s.execShellCommand(cmd)
```
Use os.write.

Stringy replied:
https://pkg.go.dev/os#example-WriteFile

Stringy reviewed (approved):
As discussed, there's a last couple of minor changes but otherwise LGTM!
